### ISSUE TITLE
panics forced to be outside of if branch

### ISSUE THREAD

**Issue by rrrkren** [onflow/cadence#1516]
### Problem

it seems if you have a panic in an else branch of a function and don't return in said branch, the contract cannot be deployed
this doesn't work and errors with `error: missing return statement`:
```
        pub fun mint(id: UInt64): @NFT {
            if id > 100 {
                return <- create NFT(initID: id)
            } else {
                panic("bad id")
            }
        }
```
this works:
```
        pub fun mint(id: UInt64): @NFT {
            if id > 100 {
                return <- create NFT(initID: id)
            } 
            panic("bad id")
        }
```

### Steps to Reproduce

https://play.onflow.org/f04d34f5-85e8-4101-8e3d-bb982d8bbd2a

**Comment by turbolent** [comment]
The return analysis only tracks definite returns and definite halts (e.g. `panic`) of branches separately, so for now the analysis is more conservative to be safe. 

It could be improved by tracking a merged status for "definitely returned or halted", which would allow this safe case.

Great suggestion!

**Comment by turbolent** [comment]
Implemented in #2090
